rmw_zenoh_cpp/src/detail/liveliness_utils.cpp

```diff
@@ -84,18 +88,136 @@ static const std::unordered_map<std::string, EntityType> str_to_entity = {
+  {std::to_string(RMW_QOS_POLICY_DURABILITY_UNKNOWN), RMW_QOS_POLICY_DURABILITY_UNKNOWN}
+};
+
 std::string zid_to_str(z_id_t id)
 {
   std::stringstream ss;
   ss << std::hex;
   size_t i = 0;
-  for (; i < (sizeof(id.id) - 1); i++) {
-    ss << static_cast<int>(id.id[i]) << ".";
+  for (; i < (sizeof(id.id)); i++) {
+    ss << static_cast<int>(id.id[i]);
   }
-  ss << static_cast<int>(id.id[i]);
   return ss.str();
 }
 
+std::vector<std::string> split_keyexpr(
+  const std::string & keyexpr,
+  const char delim = '/')
+{
+  std::vector<std::string> result = {};
+  size_t start = 0;
+  size_t end = keyexpr.find(delim);
+  while (end != std::string::npos) {
+    result.push_back(keyexpr.substr(start, end - start));
+    start = end + 1;
+    end = keyexpr.find(delim, start);
+  }
+  // Finally add the last substr.
+  result.push_back(keyexpr.substr(start));
+  return result;
+}
+
+/**
+ * Convert a rmw_qos_profile_t to a string with format:
+ *
+ * <ReliabilityKind>:<DurabilityKind>:<HistoryKind>,<HistoryDepth>"
+ * Where:
+ *  <ReliabilityKind> - enum value from rmw_qos_reliability_policy_e.
+ *  <DurabilityKind> - enum value from rmw_qos_durability_policy_e.
+ *  <HistoryKind> - enum value from rmw_qos_history_policy_e.
+ *  <HistoryDepth> - The depth number.
+ * For example, the liveliness substring for a topic with Reliability policy: reliable,
+ * Durability policy: volatile, History policy: keep_last, and depth: 10, would be
+ * "1:2:1,10". See rmw/types.h for the values of each policy enum.
+ */
+// TODO(Yadunund): Rely on maps to retrieve strings.
+std::string qos_to_keyexpr(rmw_qos_profile_t qos)
+{
+  std::string keyexpr = "";
+  keyexpr += std::to_string(qos.reliability);
+  keyexpr += QOS_DELIMITER;
+  keyexpr += std::to_string(qos.durability);
+  keyexpr += QOS_DELIMITER;
+  keyexpr += std::to_string(qos.history);
+  keyexpr += QOS_HISTORY_DELIMITER;
+  keyexpr += std::to_string(qos.depth);
+  return keyexpr;
+}
+
+/// Convert a rmw_qos_profile_t from a keyexpr. Return std::nullopt if invalid.
+std::optional<rmw_qos_profile_t> keyexpr_to_qos(const std::string & keyexpr)
+{
+  rmw_qos_profile_t qos;
+  const std::vector<std::string> parts = split_keyexpr(keyexpr, QOS_DELIMITER);
+  if (parts.size() < 3) {
+    return std::nullopt;
+  }
+  const std::vector<std::string> history_parts = split_keyexpr(parts[2], QOS_HISTORY_DELIMITER);
+  if (history_parts.size() < 2) {
+    return std::nullopt;
+  }
+
+  try {
+    qos.history = str_to_qos_history.at(history_parts[0]);
+    qos.reliability = str_to_qos_reliability.at(parts[0]);
+    qos.durability = str_to_qos_durability.at(parts[1]);
+  } catch (const std::exception & e) {
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("Error setting QoS values from strings: %s", e.what());
+    return std::nullopt;
+  }
+  // Get the history depth.
+  errno = 0;
+  char * endptr;
+  int64_t num = strtol(history_parts[1].c_str(), &endptr, 10);
```

Comment:
Given that `qos.depth` is a [size_t](https://github.com/ros2/rmw/blob/9b3d9d0e3021b7a6e75d8886e3e061a53c36c789/rmw/include/rmw/types.h#L574), I think we should use `strtoul` here:

```suggestion
  unsigned long num = strtoul(history_parts[1].c_str(), &endptr, 10);
```